cf-agent/package_module.c

```diff
@@ -0,0 +1,1611 @@
+        {
+            is_in_cache = 1;
+        }
+        else
+        {
+            Log(LOG_LEVEL_INFO,
+                "Seem to have corrupted data in cache database");
+            is_in_cache = -1;
+        }
+    }
+    
+    Log(LOG_LEVEL_DEBUG,
+        "Looking for package %s in cache returned: %d", name, is_in_cache);
+    
+    CloseDB(db_cached);
+    
+    return is_in_cache;
+}
+
+void WritePackageDataToDB(CF_DB *db_installed,
+        const char *name, const char *ver, const char *arch,
+        UpdateType type)
+{
+    char package_key[strlen(name) + strlen(ver) +
+                     strlen(arch) + 11];
+    
+    xsnprintf(package_key, sizeof(package_key),
+              "N<%s>", name);
+    if ((type == UPDATE_TYPE_UPDATES || type == UPDATE_TYPE_LOCAL_UPDATES) && 
+            HasKeyDB(db_installed, package_key, strlen(package_key) + 1))
+    {
+        size_t val_size =
+                ValueSizeDB(db_installed, package_key, strlen(package_key));
+        char buff[val_size + strlen(arch) + strlen(ver) + 8];
+        
+        ReadDB(db_installed, package_key, buff, val_size);
+        xsnprintf(buff + val_size, sizeof(package_key), "V<%s>A<%s>\n", ver, arch);
+        WriteDB(db_installed, package_key, buff, strlen(buff));
+    }
+    else if (type == UPDATE_TYPE_UPDATES || type == UPDATE_TYPE_LOCAL_UPDATES)
+    {
+        char buff[strlen(arch) + strlen(ver) + 8];
+        xsnprintf(buff, sizeof(package_key), "V<%s>A<%s>\n", ver, arch);
+        WriteDB(db_installed, package_key, buff, strlen(buff));
+    }
+    else /* UPDATE_TYPE_INSTALLED */
+    {
+        WriteDB(db_installed, package_key, "1", 1);
+        xsnprintf(package_key, sizeof(package_key),
+                "N<%s>V<%s>", name, ver);
+        WriteDB(db_installed, package_key, "1", 1);
+        xsnprintf(package_key, sizeof(package_key),
+                "N<%s>A<%s>", name, arch);
+        WriteDB(db_installed, package_key, "1", 1);
+        xsnprintf(package_key, sizeof(package_key),
+                "N<%s>V<%s>A<%s>", name, ver, arch);
+        WriteDB(db_installed, package_key, "1", 1);
+    }
+}
+
+int UpdatePackagesDB(Rlist *data, const char *pm_name, UpdateType type)
+{
+    assert(pm_name);
+    
+    bool have_error = false;
+
+    CF_DB *db_cached;
+    dbid db_id = type == UPDATE_TYPE_INSTALLED ? dbid_packages_installed :
+                                                 dbid_packages_updates;
+    if (OpenSubDB(&db_cached, db_id, pm_name))
+    {
+        CleanDB(db_cached);
+        
+        Buffer *inventory_data = BufferNewWithCapacity(INVENTORY_LIST_BUFFER_SIZE);
+
+        char *package_data[3] = {NULL, NULL, NULL};
+
+        for (const Rlist *rp = data; rp != NULL; rp = rp->next)
+        {
+            char *line = RlistScalarValue(rp);
+
+            if (StringStartsWith(line, "Name="))
+            {
+                /* We have all the information we need from previous loop
+                 * iteration. */
```

Comment:
This comment belongs inside the following block; and is misleading.  It sounds like the name we've just read goes with the data from previous iterations, where actually the name we've just read forces us to check for whether we have data from earlier iterations that we need to process before going further, since the new name starts a new record to which the old data doesn't apply.